src/apps/vpn/controller.cpp

```diff
@@ -128,6 +128,56 @@ void Controller::initialize() {
     logger.info() << "Canary Ping Succeeded";
   });
 
+  connect(SettingsHolder::instance(), &SettingsHolder::transactionBegan, this,
+          [this]() { m_wasConnectedPreTransaction = m_state == StateOn; });
+
+  connect(
+      SettingsHolder::instance(), &SettingsHolder::transactionRolledBack, this,
+      [this]() {
+        // If they are in the correct state, or going to the correct state, do
+        // nothing
+        if ((m_wasConnectedPreTransaction &&
+             (m_state == StateOn || m_state == StateConnecting ||
+              m_state == StateConfirming || m_state == StateSwitching)) ||
+            (!m_wasConnectedPreTransaction &&
+             (m_state == StateOff || m_state == StateDisconnecting)))
+          return;
+
+        if (m_wasConnectedPreTransaction) {
+          // If vpn is already off, just turn it on
+          if (m_state == StateOff) {
+            activate();
+            return;
+          }
+
+          // run lambda function in a context so we can delete it after it runs
+          // once
+          QObject* context = new QObject(this);
+          connect(this, &Controller::stateChanged, context, [this, context]() {
```

Comment:
Instead of using a boolean, you can use an enum with values, not-in-transition, connected-before-transition, disconnected-before-transition. And you don't need this `context`.